decisions/2019/09/12-tapps-workflow-iterations.md

```diff
@@ -0,0 +1,122 @@
+# 2019-09-12 @dhis2/team-apps Workflow Iteration
+
+## Attending
+
+Viktor, Hendrik, Erik, Ismay, Jan-Gerke, Birk
+
+## Automated functional tests
+
+There has been lots of talks about how we need to improve our test
+coverages to promote software quality and stability as first-class
+traits of our system.
+
+There are some trade-offs at the start while we establish the
+infrastructure for the technology and we get used to working with the
+new workflow of writing feature specifications and build functional
+tests. We consider this an investment that pays off itself.
+
+At the start of 2019 there was a cross-team initiative for how to build
+out our functional test infrastructure, and while it has lost momentum
+and focus, Cypress was the best choice.
+
+As for writing the "feature files", also known as specifications, we
+will go for the tried-and-true [Gherkin](https://cucumber.io/docs/gherkin/) specification language used by
+[Cucumber](https://cucumber.io/docs/).
+
+An example of a feature spec:
+
+```gherkin
+Feature: Show search results in hierarchy
+
+  Scenario: Show all already displayed units while search term is empty on focus
+    Given the user focuses the search input field
+    Given the user has not typed anything yet
+    Then all previously visible organisation units are still displayed
+
+  Scenario: Search result is loading
+    Given the user has entered a search term
+    When the results are being loaded
+    Then all organisation units are hidden
+    Then the loading spinner is displayed
+
+  Scenario: Some organisation units were found
+    Given the user has entered a search term
+    When some organisation units were found
+    Then the found units are shown
+    Then the unit's path is reflected in the tree structure
+    Then the unit's parents are displayed but disabled
+
+  Scenario: No organisation units were found
+    Given the user has entered a search term
+    When no organisation units were found
+    Then a text to inform the user is displayed
+```
+
+### Conclusion
+
+- [Cypress](https://www.cypress.io/) for functional tests
+- Specifications (feature files) using Gherkin
+- Regression test specs for bugs, also using Gherkin to ensure that
+  fixed bugs stay out of the system
+- Does **not** replace "other" tests like unit tests
+- Does replace Zephyr test cases (need to figure out how to integrate
+  this with QA)
+- Jan-Gerke to organise a session to demonstrate how it fits together
+
+## Better communication across the entire frontend team regarding
+libraries and shared components
```

Comment:
line-break makes it look bad in formatted view

Reply:
cefb5a6